README.md
Dotfiler: Python Development Environment
========================================

This repository contains dotfiles and it's purpose to be used as
a repository for [dotfiler][].

For convinience, don't copy these files into you homedir. Instead,
go and install [dotfiler][], then clone this repository, using
`dot add <url>` command and run `dot update` to make all necessary
symlinks.

[dotfiler]: https://github.com/svetlyak40wt/dotfiler

ChangeLog.md
## 0.7.0 (2018-03-22)

* Now ElPy installed from MELPA, not from http://jorgenschaefer.github.io/packages/
* Added M-, shortcut, to jump back to the place where M-. was called.

## 0.6.0 (2016-04-04)

* Added snippet to make a Trepan debugger set_trace.

## 0.5.0 (2016-04-01)

* Removed local key 'C-c g' for git status. Now this settings in the main config with bindings in another repository.

## 0.4.0 (2015-02-19)

* Snippets for pdb and pudb where fixed to override similar from elpy.
* Change pudb's theme to classic.

## 0.3.0 (2015-09-07)

* Don't activate yasnippets on the python level. Now it
  is activated by emacs main config directly.

## 0.2.2 (2015-04-23)

* Fixed loading of emacs config. It was broken because flymake-cursor-mode function is absent in flymake-cursor.el.

## 0.2.1 (2015-02-11)


* Fixed `column-enforce-mode` setup.

## 0.2.0

* Use `column-enforce-mode` to highlight long lines.
* Highlight not only `pdb.set_trace()`, but any line
  containing `.set_trace()`.

## 0.1.0

* Bind "C-c g" to magit-status.
* Add config and light solarized theme for [pudb][]. Color theme is compatible with iTerm2's light solarized theme.

[pudb]: https://pypi.python.org/pypi/pudb


.emacs.d/lib/python.el
;; для использования elpy, надо установить python модули:
;; pip install elpy jedi rope

(require 'package)
(add-to-list 'package-archives
             '("elpy" . "http://jorgenschaefer.github.io/packages/"))


;; elpy-enable должен быть запущен до загрузки python кода,
;; поэтому делаем это не в хуке, а прямо тут
(use-packages '(elpy))
(elpy-enable)
(setq elpy-rpc-backend "jedi")

(eval-after-load 'python
  `(progn
     (defun python-config ()
       (use-packages '(;; дополнение символов, которые используются
                       ;; в текущем файле
                       auto-complete
                       ;; подсказки неправильного синтаксиса и импортов
                       flymake-python-pyflakes
                       ;; этот плагин тормозил
                       ;; flymake-cursor
                       ;; выделение окружающего контекста по C-c =
                       expand-region
                       ;; выделение строк, которые выбиваются за
                       ;; границу в 80 символов
                       column-enforce-mode
                      ))

       ;; To make right indent after if/for/etc tokens
       (local-set-key (kbd "RET") 'newline-and-indent)
       ;; Nothing on this shortcut
       (local-set-key (kbd "C-c n") 'flymake-goto-next-error)
       (local-set-key (kbd "C-c i") 'iedit-mode)

       ;; To delete trailing whitespaces on save
       ;; This hook was commented out, because on file with 5000 lines,
       ;; deletion takes about 5 seconds each time when I'm saving the file.
       ;; (add-hook 'before-save-hook 'delete-trailing-whitespace nil t)
  
       ;; подсвечиваем брекпоинты, чтобы не забыть их удалить
       ;; это включает hi-lock minor mode
       (highlight-lines-matching-regexp "\.set_trace" "hi-blue")

       ;; Turn on automatic syntax checker
       ;; https://github.com/purcell/flymake-python-pyflakes
       (require 'flymake-python-pyflakes)
       (flymake-python-pyflakes-load)
       (column-enforce-mode))
     
     (add-hook 'python-mode-hook 'python-config)))




.config/pudb/solarized-light.py
# coding: utf-8

# Slightly modified "solarized" theme for pudb
# ============================================
#
# Supported 16 color values:
#   'h0' (color number 0) through 'h15' (color number 15)
#    or
#   'default' (use the terminal's default foreground),
#   'black', 'dark red', 'light green', 'brown', 'dark blue',
#   'dark magenta', 'dark cyan', 'light gray', 'dark gray',
#   'light red', 'light green', 'yellow', 'light blue',
#   'light magenta', 'light cyan', 'white'
#
# Supported 256 color values:
#   'h0' (color number 0) through 'h255' (color number 255)
#
# 256 color chart: http://en.wikipedia.org/wiki/File:Xterm_color_chart.png
#
# Based on XCode's midnight theme
# Looks best in iTerm2 with light solarized color theme.
#
# ATTENTION: Be sure to set 'raw' terminal driver in pudb's settings
#            to enable 256 color support!

palette.update({
    # UI
    "header": ("black", "light blue", "standout"),
    "focused sidebar": ("yellow", "light blue", "standout"),
    "group head": ("black", "light blue"),
    "background": ("black", "light blue"),
    "label": ("black", "light blue"),
    "value": ("white", "dark blue"),
    "fixed value": ("black", "light blue"),

    "variables": ("light blue", "default"),

    "var label": ("dark blue", "default"),
    "var value": ("light blue", "default"),

    "focused var label": ("white", "dark blue"),
    "focused var value": ("black", "dark blue"),

    "highlighted var label": ("white", "light green"),
    "highlighted var value": ("white", "light green"),
    "focused highlighted var label": ("white", "light green"),
    "focused highlighted var value": ("white", "light green"),

    "stack": ("light blue", "default"),

    "frame name": ("dark blue", "default"),
    "frame class": ("light blue", "default"),
    "frame location": ("light green", "default"),

    "focused frame name": ("white", "dark blue"),
    "focused frame class": ("black", "dark blue"),
    "focused frame location": ("dark gray", "dark blue"),

    "focused current frame name": ("white", "light green"),
    "focused current frame class": ("black", "light green"),
    "focused current frame location": ("dark gray", "light green"),

    "current frame name": ("white", "light green"),
    "current frame class": ("black", "light green"),
    "current frame location": ("dark gray", "light green"),

    # breakpoints
    "breakpoint": ("light blue", "default"),
    "focused breakpoint": ("white", "light green"),
    "current breakpoint": ("white", "dark blue"),
    "focused current breakpoint": ("white", "dark blue"),

    # source
    "breakpoint source": ("light blue", "black"),
    "current breakpoint source": ("black", "light green"),
    "breakpoint focused source": ("dark gray", "dark blue"),
    "current breakpoint focused source": ("black", "light green"),
    "breakpoint marker": ("dark red", "default"),

    "search box": ("default", "default"),

    "source": ("light blue", "default"),
    "current source": ("h231", "h211"),
    "current focused source": ("h231", "h140"),

    "focused source": ("dark gray", "h69"),

    "current highlighted source": ("black", "dark cyan"),
    "highlighted source": ("light blue", "black"),

    "line number": ("light blue", "default"),
    "keyword": ("dark green", "default"),
    "name": ("light blue", "default"),
    "literal": ("dark cyan", "default"),
    "string": ("dark cyan", "default"),
    "doublestring": ("dark cyan", "default"),
    "singlestring": ("light blue", "default"),
    "docstring": ("dark cyan", "default"),
    "backtick": ("light green", "default"),
    "punctuation": ("light blue", "default"),
    "comment": ("light green", "default"),
    "classname": ("dark blue", "default"),
    "funcname": ("dark blue", "default"),

    # shell

    "command line edit": ("light blue", "default"),
    "command line prompt": ("light blue", "default"),

    "command line output": ("light blue", "default"),
    "command line input": ("light blue", "default"),
    "command line error": ("dark red", "default"),

    "focused command line output": ("black", "light green"),
    "focused command line input": ("black", "light green"),
    "focused command line error": ("dark red", "light blue"),

    "command line clear button": ("light blue", "default"),
    "command line focused button": ("black", "light blue"),
})


.emacs.d/lib/50-python.el
;; для использования elpy, надо установить python модули:
;; pip install elpy jedi rope

(require 'package)

;; Начиная с версии 1.18.0 elpy стали раздавать через MELPA stable.
;; (add-to-list 'package-archives
;;              '("elpy" . "http://jorgenschaefer.github.io/packages/"))


;; elpy-enable должен быть запущен до загрузки python кода,
;; поэтому делаем это не в хуке, а прямо тут
(use-packages '(elpy))
(elpy-enable)
(setq elpy-rpc-backend "jedi")

 
(defun reformat-test-name (&optional arg)
  "Эта функция превращает строку с описанием теста типа
test_set_signature (test.test_settings.TestSettings)
в
test.test_settings:TestSettings.test_set_signature
"
  (interactive "p")
  (kmacro-exec-ring-item
   (quote ([67108896 19 32 return left 24 11 5 backspace 46 25 18 92 46 18 return 4 58 1 4 4 14] 0 "%d"))
   arg))


;; (when (executable-find "ipython")
;;   (setq python-shell-interpreter "ipython")
;;   (setq python-shell-interpreter-args "--pylab"))

(eval-after-load 'python
  `(progn
     (defun python-config ()
       (use-packages '(;; дополнение символов, которые используются
                       ;; в текущем файле
                       auto-complete
                       ;; подсказки неправильного синтаксиса и импортов
                       flymake-python-pyflakes
                       ;; этот плагин тормозил
                       ;; flymake-cursor
                       ;; выделение окружающего контекста по C-c =
                       expand-region
                       ;; выделение строк, которые выбиваются за
                       ;; границу в 80 символов
                       column-enforce-mode
                      ))

       ;; To make right indent after if/for/etc tokens
       (local-set-key (kbd "RET") 'newline-and-indent)
       ;; Nothing on this shortcut
       (local-set-key (kbd "C-c n") 'flymake-goto-next-error)
       (local-set-key (kbd "C-c i") 'iedit-mode)

       ;; Чтобы прыгать обратно на место откуда делался поиск тега, как в Lisp
       ;; по умолчанию, это было забинжено на M-*, что неудобно нажимать
       ;; А на M-, по умолчанию висело tags-loop-continue, что я никогда не исользую.
       (local-set-key (kbd "M-,") 'pop-tag-mark)

       ;; To delete trailing whitespaces on save
       ;; This hook was commented out, because on file with 5000 lines,
       ;; deletion takes about 5 seconds each time when I'm saving the file.
       ;; (add-hook 'before-save-hook 'delete-trailing-whitespace nil t)
  
       ;; подсвечиваем брекпоинты, чтобы не забыть их удалить
       ;; это включает hi-lock minor mode
       (highlight-lines-matching-regexp "\.set_trace" "hi-blue")

       ;; Turn on automatic syntax checker
       ;; https://github.com/purcell/flymake-python-pyflakes
       (require 'flymake-python-pyflakes)
       (flymake-python-pyflakes-load)
       (column-enforce-mode))
     
     (add-hook 'python-mode-hook 'python-config)))


